compiler/rustc_errors/src/lib.rs

```diff
@@ -1295,20 +1299,15 @@ impl DiagCtxtInner {
                 return None;
             }
             Warning if !self.flags.can_emit_warnings => {
-                if diagnostic.has_future_breakage() {
-                    (*TRACK_DIAGNOSTIC)(diagnostic, &mut |_| {});
-                }
                 return None;
             }
             Allow | Expect(_) => {
-                (*TRACK_DIAGNOSTIC)(diagnostic, &mut |_| {});
                 return None;
```

Comment:
Is this correct for `Expect` diagnostics too?
The second execution is not really a no-op, as this function will add a side-effect to `self.fulfilled_expectations`.
I'm not certain that we correctly test this path in incremental.

Reply:
Hmm, I think you are right. In fact, similar reasoning applies for the `Warning` case -- there is the side-effect of `self.future_breakage_diagnostics` being updated on the replay. So neither call can be removed.

In fact, now I suspect there is a missing `TRACK_DIAGNOSTIC` call for this code:
```
            if let LintExpectationId::Unstable { .. } = expectation_id {
                self.unstable_expect_diagnostics.push(diagnostic);
                return None;
            }
```
because there's the side-effect of updating `self.unstable_expect_diagnostics`.

Reply:
And now I'm wondering if there's a missing `TRACK_DIAGNOSTIC` for the delayed bugs as well :(

Reply:
Are you sure about the delayed bugs case? They don't doom compilation unless no other errors are emitted. So if we do diagnostic replay and don't reply delayed bugs, I think we are possibly missing out on some internal sanity checking.